irr/include/SSkinMeshBuffer.h

```diff
@@ -7,12 +7,153 @@
 #include "IMeshBuffer.h"
 #include "CVertexBuffer.h"
 #include "CIndexBuffer.h"
+#include "IVertexBuffer.h"
 #include "S3DVertex.h"
+#include "vector3d.h"
+#include <algorithm>
 #include <cassert>
+#include <memory>
+#include <numeric>
+#include <optional>
 
 namespace scene
 {
 
+struct WeightBuffer
+{
+	constexpr static u16 MAX_WEIGHTS_PER_VERTEX = 4;
+	size_t n_verts;
+	std::unique_ptr<u16[]> joint_idxs;
+	std::unique_ptr<f32[]> weights;
+
+	std::optional<std::vector<u32>> animated_vertices;
+
+	// A bit of a hack for now: Store static positions here so we can use them for skinning.
+	// Ideally we might want a design where we do not mutate the original vertex buffer at all.
+	std::unique_ptr<core::vector3df[]> static_positions;
+	std::unique_ptr<core::vector3df[]> static_normals;
+
+	WeightBuffer(size_t n_verts) :
+		n_verts(n_verts),
+		joint_idxs(std::make_unique<u16[]>(MAX_WEIGHTS_PER_VERTEX * n_verts)),
+		weights(std::make_unique<f32[]>(MAX_WEIGHTS_PER_VERTEX * n_verts))
+	{}
+
+	u16 *getJointIndices(u32 vertex_id)
+	{ return &joint_idxs[vertex_id * MAX_WEIGHTS_PER_VERTEX]; }
+	const u16 *getJointIndices(u32 vertex_id) const
+	{ return &joint_idxs[vertex_id * MAX_WEIGHTS_PER_VERTEX]; }
+
+	f32 *getWeights(u32 vertex_id)
+	{ return &weights[vertex_id * MAX_WEIGHTS_PER_VERTEX]; }
+	const f32 *getWeights(u32 vertex_id) const
+	{ return &weights[vertex_id * MAX_WEIGHTS_PER_VERTEX]; }
+
+	void addWeight(u32 vertex_id, u16 joint_id, f32 weight)
+	{
+		assert(vertex_id < n_verts);
+		assert(weight >= 0.f);
+		f32 *weights = getWeights(vertex_id);
+		f32 *min_weight = std::min_element(weights, weights + MAX_WEIGHTS_PER_VERTEX);
+		if (*min_weight > weight)
+			return;
+
+		*min_weight = weight;
+		getJointIndices(vertex_id)[min_weight - weights] = joint_id;
+	}
+
+	void skinVertex(u32 vertex_id, core::vector3df &pos, core::vector3df &normal,
+			const std::vector<core::matrix4> &joint_transforms) const
+	{
+		const u16 *joint_idxs = getJointIndices(vertex_id);
```

Comment:
Might it make sense to move these implementations to a new source file? - or is this header required by server builds too?

Reply:
moved `WeightBuffer` to its own header & source file:

https://github.com/luanti-org/luanti/pull/16655/commits/61c77cb0a45f85c2c310ed4df121c84c440467ef